insonmnia/miner/server.go

```diff
@@ -80,6 +83,49 @@ func (m *Miner) Handshake(context.Context, *pb.HandshakeRequest) (*pb.HandshakeR
 	return nil, status.Errorf(codes.Aborted, "not implemented")
 }
 
+func (m *Miner) scheduleStatusPurge(id string) {
+	t := time.NewTicker(time.Second * 30)
+	defer t.Stop()
+	for {
+		select {
+		case <-t.C:
+			m.mu.Lock()
+			delete(m.containers, id)
+			m.mu.Unlock()
+		case <-m.ctx.Done():
+			return
+		}
+	}
+
+}
+
+func (m *Miner) setStatus(status *pb.TaskStatus, id string) {
+	m.mu.Lock()
+	_, ok := m.containers[id]
+	if !ok {
+		m.containers[id] = &ContainerInfo{}
+	}
+	m.containers[id].status = status
+	if status.Status == pb.TaskStatus_BROKEN || status.Status == pb.TaskStatus_FINISHED {
+		go m.scheduleStatusPurge(id)
+	}
+	for _, ch := range m.statusChannels {
+		ch <- true
```

Comment:
add select ctx.Done(). Otherwise it might lead to a deadlock if sendUpdatesOnNotify has quited before